File: leetcode/CPP/179-Largest-Number.cpp
/*
179. Largest Number
给定一组非负整数，重新排列它们的顺序使之组成一个最大的整数
*/
bool comp(int num1, int num2)
{
    string s1 = to_string(num1) + to_string(num2);
    string s2 = to_string(num2) + to_string(num1);
    return s1 > s2;
}
string smallestNumber(vector<int> &nums)
{
    sort(nums.begin(), nums.end(), comp);
    if (!nums.empty() && nums[0] == 0)
        return "0";
    string res = "";

    for (int num : nums)
        res = res + to_string(num);
    return res;
}

File: leetcode/python/179-Largest-Number.py
'''
179. Largest Number
Given a list of non negative integers, arrange them such that they 
form the largest number.
Input: [10,2]
Output: "210"
'''
# sort按小于排序，如果你认为第一个参数“小于”第二个参数（返回值<0)，则sort会把它放前面
from functools import cmp_to_key
def smallestNumber(nums):
    if any(nums):
        nums = sorted([str(x) for x in nums], key=cmp_to_key(lambda x, y: int(x + y) - int(y + x)), reverse=True)
        return ''.join(nums)
    return '0'